lib/vm/src/instance/mod.rs

```diff
@@ -776,6 +782,205 @@ impl Instance {
             self.imported_table(table_index).handle
         }
     }
+
+    // To implement Wait / Notify, a HasMap, behind a mutex, will be used
+    // to track the address of waiter. The key of the hashmap is based on the memory
+    // and waiter threads are "park"'d (with or without timeout)
+    // Notify will wake the waiters by simply "unpark" the thread
+    // as the Thread info is stored on the HashMap
+    // once unparked, the waiter thread will remove it's mark on the HashMap
+    // timeout / awake is tracked with a boolean in the HashMap
+    // because `park_timeout` doesn't gives any information on why it returns
+    fn do_wait(&mut self, index: u32, dst: u32, timeout: i64) -> u32 {
+        // fetch the notifier
+        let key = (index, dst);
+        let mut conds = self.conditions.lock().unwrap();
+        conds.entry(key).or_insert_with(Vec::new);
+        let v = conds.get_mut(&key).unwrap();
+        v.push((current(), false));
```

Comment:
Yeah, I think a test is overkill.

It's still good to have the limit for Deploy, because otherwise a single deployment could eat up all the memory by just spawning lots of waiter threads in a loop.

Well, we'll actually we'll always have max thread limits, so that's also not a huge problem.
Still doesn't hurt to adhere to the spec, `HashMap::len()` is not expensive.